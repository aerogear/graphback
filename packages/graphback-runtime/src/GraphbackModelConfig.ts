
/**
 * Interface represents additional metadata about a model
 * that may change behaviour at runtime
 */
export interface GraphbackModelConfig {
    name: string;
    pubSub: PubSubConfig;
}

/**
 * Configuration for the publish subscribe model
 */
export interface PubSubConfig {
    publishCreate?: boolean;
    publishUpdate?: boolean;
    publishDelete?: boolean;
}

