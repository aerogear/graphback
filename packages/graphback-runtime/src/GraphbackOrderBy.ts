type SortDirection = 'asc' | 'desc'

/**
 * @todo move this into a common file
 */
export interface GraphbackOrderBy {
  direction?: SortDirection
  field: string
}
