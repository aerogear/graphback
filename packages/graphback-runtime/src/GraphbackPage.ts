/**
 * Interface represents GraphbackPage type
 */
export interface GraphbackPage {
    limit?: number,
    offset?: number,
};