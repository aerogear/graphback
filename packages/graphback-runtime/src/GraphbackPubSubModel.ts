
/**
 * Interface represents graphback models that are generated by default resolver generator
 */
export interface GraphbackPubSubModel {
  name: string,
  pubSub: PubSubConfig
}

/**
 * Configuration for the publish subscribe model
 */
export interface PubSubConfig {
  publishCreate?: boolean;
  publishUpdate?: boolean;
  publishDelete?: boolean;
}

