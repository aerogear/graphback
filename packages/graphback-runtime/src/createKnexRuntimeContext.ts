import { PubSubEngine } from 'graphql-subscriptions';
import * as Knex from 'knex'
import { CRUDService, PgKnexDBDataProvider } from '.';
import { GraphbackRuntimeContext } from './GraphbackRuntimeContext';
 
/**
 * Create context object required for the graphback runtime layer 
 */
export const createKnexRuntimeContext = (db: Knex, pubSub: PubSubEngine): GraphbackRuntimeContext => {
  const crudDb = new PgKnexDBDataProvider(db);
  const crudService = new CRUDService(crudDb, pubSub);

  return {
    crudService,
    crudDb,
    pubSub
  };
}
