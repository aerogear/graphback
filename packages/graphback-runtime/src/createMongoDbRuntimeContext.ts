import { GraphQLSchema } from 'graphql';
import { PubSubEngine } from 'graphql-subscriptions';

/**
 * Create context object required for the graphback runtime layer 
 */
export const createMongoRuntimeContext = (db: any, pubSub: PubSubEngine, schema: GraphQLSchema) => {
  return {}
}
