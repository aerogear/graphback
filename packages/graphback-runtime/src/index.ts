
// DB
export * from "./data/GraphbackDataProvider"
export * from "./data/NoDataError"
export * from "./data/PgKnexDBDataProvider"
export * from "./data/KnexDBDataProvider"

// Service
export * from "./service/CRUDService"
export * from "./service/GraphbackCRUDService"

export * from "./resolvers/LayeredRuntimeResolverGen"
export * from "./resolvers/RuntimeResolversDefinition"

// Helpers 
export * from "./GraphbackRuntimeContext"
export * from "./createKnexRuntimeContext"
export * from "./validateRuntimeContext"