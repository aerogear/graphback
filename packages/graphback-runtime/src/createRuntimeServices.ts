import { GraphQLObjectType, GraphQLSchema } from "graphql"
import { PubSubEngine } from 'graphql-subscriptions';
import { GraphbackDataProvider } from './data/GraphbackDataProvider';
import { CRUDService } from './service/CRUDService';
import { GraphbackModelConfig } from "./GraphbackModelConfig";

export type RuntimeServicesOptions<Provider extends GraphbackDataProvider, Service extends CRUDService> = {
  schema: GraphQLSchema,
  models: GraphbackModelConfig[]
  db: any
  pubSub: PubSubEngine
  provider: new (...args: any[]) => Provider,
  service: new (...args: any[]) => Service;
};

/**
 * Helper function for creating array of datasources based on the model files that will 
 * be able to connect to Postgress database using knex
 * 
 * @param schema 
 * @param db 
 * @param pubSub 
 */
export function createRuntimeServices<Provider extends GraphbackDataProvider, Service extends CRUDService> 
  (servicesConfig: RuntimeServicesOptions<Provider, Service>) {

  const { models, schema, db, pubSub, provider, service } = servicesConfig
  if (!models || models.length === 0) {
    throw new Error(`No models provided`)
  }

  return models.reduce((services: any, model: GraphbackModelConfig) => {
    const modelType = schema.getType(model.name) as GraphQLObjectType
    if (modelType === undefined) {
      throw new Error(`
      Schema is missing provided type. 
      Please make sure that you pass the right schema to createCRUDRuntimeContext`)
    }

    const objectDB = new provider(modelType, db)

    services[model.name] = new service(modelType, objectDB, {
      pubSub,
      ...model.pubSub
    })

    return services;
  }, {})
}