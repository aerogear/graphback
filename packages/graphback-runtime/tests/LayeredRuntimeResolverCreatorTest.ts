//tslint:disable-next-line: match-default-export-name no-implicit-dependencies
import { buildSchema } from 'graphql';
import { GraphbackCoreMetadata } from '@graphback/core';
import { LayeredRuntimeResolverCreator } from '../src';

// @ts-ignore
// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
// const test = _test as TestInterface<{}>;

const schema = buildSchema(`
"""
@model
"""
type Todos {
 id: ID!
 text: String 
}

"""
@model
"""
type User {
  name: String
  """
  @db.primary
  """
  username: String
}
`);


test('find Todo by text', async ()  => {
  const metadata = new GraphbackCoreMetadata({ crudMethods: {} }, schema)
  const generator = new LayeredRuntimeResolverCreator(metadata.getModelDefinitions(), {});
  expect(generator.generate()).toMatchSnapshot();
});

