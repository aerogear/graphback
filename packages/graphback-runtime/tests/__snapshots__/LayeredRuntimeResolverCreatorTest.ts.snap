// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`find Todo by text 1`] = `
Object {
  "Mutation": Object {
    "createTodos": [Function],
    "createUser": [Function],
    "deleteTodos": [Function],
    "deleteUser": [Function],
    "updateTodos": [Function],
    "updateUser": [Function],
  },
  "Query": Object {
    "findTodos": [Function],
    "findUsers": [Function],
  },
  "Subscription": Object {
    "deletedTodos": Object {
      "subscribe": [Function],
    },
    "deletedUser": Object {
      "subscribe": [Function],
    },
    "newTodos": Object {
      "subscribe": [Function],
    },
    "newUser": Object {
      "subscribe": [Function],
    },
    "updatedTodos": Object {
      "subscribe": [Function],
    },
    "updatedUser": Object {
      "subscribe": [Function],
    },
  },
}
`;
